src/main/kotlin/org/rust/ide/refactoring/move/common/RsMoveConflictsDetector.kt

```diff
@@ -0,0 +1,159 @@
+        val itemInTempMod = tempMod.addInner(item)
+        val targetInTempMod = itemInTempMod.descendantsOfType<RsVisible>()
+            .singleOrNull { it.getCopyableUserData(RS_ELEMENT_FOR_CHECK_INSIDE_REFERENCES_VISIBILITY) == target }
+            ?: return
+
+        if (!targetInTempMod.isVisibleFrom(referenceElement.containingMod)) {
+            addVisibilityConflict(conflicts, referenceElement, target)
+        }
+        target.putCopyableUserData(RS_ELEMENT_FOR_CHECK_INSIDE_REFERENCES_VISIBILITY, null)
+    }
+
+    private fun detectPrivateFieldOrMethodOutsideReferences() {
+        fun checkVisibility(referenceElement: RsElement, target: RsVisible) {
+            if (!target.isInsideMovedElements(elementsToMove) && !target.isVisibleFrom(targetMod)) {
+                addVisibilityConflict(conflicts, referenceElement, target)
+            }
+        }
+
+        val elementsToCheck = movedElementsDeepDescendantsOfType<RsElement>(elementsToMove)
+        detectPrivateFieldOrMethodReferences(elementsToCheck, ::checkVisibility)
+    }
+
+    private fun detectPrivateFieldOrMethodReferences(
+        elementsToCheck: Sequence<RsElement>,
+        checkVisibility: (RsElement, RsVisible) -> Unit
+    ) {
+        fun checkVisibility(reference: RsReferenceElement) {
+            val target = reference.reference?.resolve() as? RsVisible ?: return
+            checkVisibility(reference, target)
+        }
+
+        loop@ for (element in elementsToCheck) {
+            when (element) {
+                is RsDotExpr -> {
+                    val fieldReference = element.fieldLookup ?: element.methodCall ?: continue@loop
+                    checkVisibility(fieldReference)
+                }
+                is RsStructLiteralField -> {
+                    val field = element.resolveToDeclaration() ?: continue@loop
+                    checkVisibility(element, field)
+                }
+                is RsPatField -> {
+                    val patBinding = element.patBinding ?: continue@loop
+                    checkVisibility(patBinding)
+                }
+                is RsPatTupleStruct -> {
+                    // it is ok to use `resolve` and not `deepResolve` here
+                    // because type aliases can't be used in destructuring tuple struct:
+                    val struct = element.path.reference?.resolve() as? RsStructItem ?: continue@loop
+                    val fields = struct.tupleFields?.tupleFieldDeclList ?: continue@loop
+                    for (field in fields) {
+                        checkVisibility(element, field)
+                    }
+                }
+                is RsPath -> {
+                    // conflicts for simple paths are handled using `pathNewAccessible`/`pathNewFallback` machinery
+                    val isInsideSimplePath = element.parentsWithSelf
+                        .takeWhile { it is RsPath }
+                        .any { isSimplePath(it as RsPath) }
```

Comment:
Not sure I undestand this code right. Does `.any` should be really used, and not `.all`? Or maybe even `isSimplePath(element.rootPath())`?

Reply:
Any outermost path has exactly one simple subpath (please see updated comment for `isSimplePath`), so we can't use `all` because it will almost always return `false`.

---

Consider this example: `mod1::mod2::Struct1::func();`. Here the only simple path is `mod1::mod2::Struct1`. It is not necessary to check visibility for path `mod1::mod2`, because we already checked visibility for `mod1`, `mod2` (and `Struct1`) when searched `pathNewAccessbile` using `RsImportHelper`. But we need to check visibility for path `mod1::mod2::Struct1::func` (that is for `func`)

Reply:
Thank you, I got it. Probably `isSimplePath` is not the best name, but I don't have any good variants right now. Please add a test (if there is no such) that will fail when `.all` or `isSimplePath(element.rootPath())` is used instead to protect this code from such breaking changes

Reply:
Ok, thank you, I added tests